98 dbt cloud ide support

owlas opened the pull request:
Closes #197 

Updated network error handling

## Testing we get actionable error messages

### Local

- [x] server fails to start
- [x] project dir not found
- [x] profiles dir not found
- [x] profile not found in profiles.yml
- [x] bigquery oauth but no gcloud config
- [x] server can't compile


### Cloud

- [x] Api key not valid
- [x] Wrong resource (account, environment)
- [x] IDE not active
- [x] Project doesn't compile
- [x] Lightdash syntax doesn't compile

ZeRego commented on lightdash.yml:
```diff
@@ -2,6 +2,7 @@ version: '1.0'
-    type: 'dbt'
-    profiles_dir: /usr/app/profiles
-    project_dir: /usr/app/dbt
+    type: 'dbt_cloud_ide'
+    account_id: '20983'
+    environment_id: '32747'
+    project_id: '34669'
```
Should this be public ?

owlas replied:
Yah it's ok

ZeRego commented on packages/backend/src/dbt/dbtRpcClient.ts:
```diff
@@ -3,64 +3,15 @@ import fetch from 'node-fetch';
-        return data;
-    }
+export abstract class DbtRpcClientBase {
+    abstract _postDbtSyncRpc(
```
If this returns a promise, shouldn't we call it async ? we might need to rethink this name since we already have a `_postDbtAsyncRpc`

owlas replied:
Yep that's really confusing, it's meant to refer to how the procedure runs over http not a reference to the js function. But I'll change this

ZeRego commented on packages/backend/src/projectAdapters/projectAdapter.ts:
```diff
@@ -2,10 +2,14 @@ import { DbtProjectConfig, ProjectType } from '../config/parseConfig';
 
 export const projectAdapterFromConfig = (
     config: DbtProjectConfig,
+    // eslint-disable-next-line consistent-return
 ): ProjectAdapter => {
+    console.log(config);
+    // eslint-disable-next-line default-case
```
fyi: These two rules can be solved by returning an error on switch default case.

owlas replied:
fixed

ZeRego commented on packages/backend/src/config/parseConfig.ts:
```diff
@@ -27,7 +28,18 @@ interface DbtRemoteProjectConfig extends DbtProjectConfigBase {
 }
 
-export type DbtProjectConfig = DbtLocalProjectConfig | DbtRemoteProjectConfig;
+interface DbtCloudIDEProjectConfig extends DbtProjectConfigBase {
```
note: don't forget to update `mergeWithEnvironment()` logic

owlas commented:
@ZeRego updated: 

- Refactored the rpc client to be more strict on types received from dbt (i.e. error and results types)
- Improved error handling when receiving unknown and known error responses (from cloud + local)
- Renamed methods to make more sense
- Simplified polling and reduced timeout length for waiting for server start

owlas commented on packages/backend/src/projectAdapters/projectAdapter.ts:
```diff
@@ -18,7 +20,15 @@ export const projectAdapterFromConfig = (
         default:
-            throw new Error('Project adapter not implemented');
+            const never: never = config;
+            throw new Error('NOPE!');
```
update

owlas commented on packages/backend/src/lightdash.ts:
```diff
@@ -7,6 +7,7 @@ import { lightdashConfig } from './config/lightdashConfig';
 // Setup dbt adapter
 const projectConfig = lightdashConfig.projects[0];
 const adapter = projectAdapterFromConfig(projectConfig);
+console.log(`HELLO ITS AN ADAPTER ${adapter}`);
```
debug line

owlas commented on packages/backend/src/errors.ts:
```diff
@@ -109,8 +95,18 @@ export class DbtError extends LightdashError {
+        super({
+            message,
+            name: 'RetryableNetworkError',
+            statusCode: 503, // update this status code?
```
fix

owlas commented:
closes #197

ZeRego commented on packages/backend/src/dbt/dbtRpcClientBase.ts:
```diff
@@ -0,0 +1,291 @@
+                ),
+            );
+        }
+        console.log(results);
```
Forgotten console.log

ZeRego commented on packages/backend/src/dbt/dbtRpcClientBase.ts:
```diff
@@ -0,0 +1,291 @@
+        }
+        attempts += 1;
+        interval = Math.min(maxInterval, interval * intervalMultiplier);
+        setTimeout(poll, interval, resolve, reject);
```
We should extract the retry mechanism into a util function. 
I found this answer and it looks clean, what you think ? 
https://stackoverflow.com/a/44577075

owlas replied:
I thought this was pretty barebones. It's reusable retry that works with Lightdash specific Errors. 

What two functions would you break this into?

ZeRego replied:
It was in case we wanted to use a generic promise retry in multiple places beside connecting to dbt server. 
We don't have to change this now and we can re-evaluate later if we need to.

ZeRego commented on packages/backend/src/dbt/dbtCloudV2RpcClient.ts:
```diff
@@ -0,0 +1,78 @@
+        } catch (e) {
+            if (e instanceof NetworkError) {
+                // handle common dbt Cloud expected responses
+                console.log(e.data);
```
Forgotten console.log

ZeRego reviewed (approved):
LGTM 🥳 

We just need to remove the console.logs

owlas commented:
@ZeRego updated all comments and bumped support to 0.20.0

We now require 0.20.0 (I think it's good to make this change asap).

We can't get the version number from the rpc server (we can hack it locally by reading stdout).

Instead I've tried to make it very obvious in the documentation that lightdash requires dbt 0.20.0

owlas commented on packages/backend/src/dbt/translator.ts:
```diff
@@ -19,7 +19,21 @@ import { DepGraph } from 'dependency-graph';
+    const [project, ...rest] = patchPath.split('://');
+    if (rest.length === 0) {
+        throw new DbtError(
+            'Could not parse dbt manifest. It looks like you might be usig an old version of dbt. You must be using dbt version 0.20.0 or above.',
```
typo

ZeRego commented on docs/docs/get-started/setup-a-dbt-cloud-project.md:
```diff
@@ -0,0 +1,107 @@
+projects:
+  - name: 'default'
+    type: 'dbt_cloud_ide'
+    accountId: # your id
+    projectId: # your id
+    environmentId: # your id
+    apiKey: # your api key
```
Amend example
```suggestion
    account_id: # your id
    project_id: # your id
    environment_id: # your id
    api_key: # your api key
```
After amending this locally I also got the parsing error `ParseError: Lightdash config file successfully loaded but invalid: Field at /projects/0/account_id must be string` because the values are by default numbers. So I had to wrap them in single quotes. Should we mention that in the example or change or validation to also support numbers ?

owlas replied:
Not sure what to do here, the ids are numbers. But always feels weird to make ids an actual number haha.

What would you prefer? Accepting numbers and strings doesn't seem too bad

owlas replied:
String already means people can type any invalid id in there that doesn't look like a number

owlas replied:
Oh one thing to add: dbt cloud returns a different response in these cases:
1.  if you give a valid id but doesn't exist like '999999' - you get a meaningful response
2. you give a weird id like 'aaa11happy22no' - you get a garbage response

ZeRego replied:
> dbt cloud returns a different response
We should open a bug on their repo. I don't see the reason why they are treated differently.

> Accepting numbers and strings doesn't seem too bad
Yeah, I think this is the best approach

owlas replied:
> should open a bug on their repo

I think this is proprietary (i.e. not in dbt core)

> Yeah, I think this is the best approach

on it!

owlas commented:
@ZeRego just bumped some docs changes otherwise I think this is good to go
